gprofiler/main.py

```diff
@@ -137,12 +138,10 @@ def __init__(
         # the latter can be root only. the former can not. we should do this separation so we don't expose
         # files unnecessarily.
         self._temp_storage_dir = TemporaryDirectoryWithMode(dir=TEMPORARY_STORAGE_PATH, mode=0o755)
-        self.system_profiler, self.process_profilers = get_profilers(
-            user_args,
-            storage_dir=self._temp_storage_dir.name,
-            stop_event=self._stop_event,
-            profile_spawned_processes=self._profile_spawned_processes,
+        self._profiler_state = ProfilerState(
+            self._stop_event, self._temp_storage_dir.name, self._profile_spawned_processes
```

Comment:
It's nicer if now `GProfiler` uses the fields from `_profiler_state`, i.e - don't store `_stop_event`, `_temp_storage_dir`, `_profile_spawned_processes` on `GProfiler` - but only on the `State` class.

You can pass the `TemporaryDirectoryWithMode(dir=TEMPORARY_STORAGE_PATH, mode=0o755)` as a parameter to the state class, the expose the storage dir with a `@property`?

Reply:
I did it a bit differently. Since we pass string in tests ProfilerState constructor should accept string as storage_dir too.

Reply:
Hmm I commented on that later - https://github.com/Granulate/gprofiler/pull/668#discussion_r1096231221 - it is not a good idea to make the tested code more complex to acommodate tests, instead of makes more sense to make the tests code a bit more complex. In that case, it's just calling the `TemporaryStorageDirectory` on the base path `tmp_path` from pytest, right? I think it's a better solution.

Alternatively - you can move the creation of `TemporaryStorageDirectory` inside `ProfilerState`, then both the tests and `GProfiler` will pass it a `str`. Up to you.